[KZL-761] Add Golang Getting Started page (#147)
kuzzleio/documentation-V2: src/sdk-reference/go/1/getting-started/index.md

---
layout: sdk.html.hbs
algolia: true
title: Getting started
description: Getting started
order: 100
separator: essentials
---

# Getting Started

In this tutorial you will learn how to install the Kuzzle **Go SDK**.
We will walk you through creating scripts that can **store** documents in Kuzzle and subscribe to **notification** for each new document created.

Before proceeding, please make sure your system has these programs installed:

* **Go** version 1.9 or higher (<a href="https://golang.org/doc/install">instructions here</a>)
* A running Kuzzle server.

## Installation

You can easily install our Go SDK:

```bash
$ go get github.com/kuzzleio/sdk-go
```

This will fetch the SDK and install it in your `GOPATH`.

## First connection

> Initialize a new Go project as describe in [Go Documentation](https://golang.org/doc/code.html#Command).

Reviewer: ```suggestion
Initialize a new Go project as described in [Go Documentation](https://golang.org/doc/code.html#Command).
```

> Then, create a `init.go` file and start by adding the code below:

Reviewer: ```suggestion
Then create a `init.go` file and start by adding the code below:
```

[snippet=init]

This program will initialize our Kuzzle creating an index and a collection
we'll use in the next section.
Run this little bunch of code:

```bash
$ go run init.go
Connected!
Index nyc-open-data created!
Collection yellow-taxi created!
```

Congratulations, you performed a first connection to Kuzzle with a Go program.
You are now able to:
* Load the `Kuzzle Go SDK` from your `GOPATH`.
* Instantiate a protocol (here `websocket`) and a Kuzzle SDK instance.
* Connect to Kuzzle instance running on `localhost` with the `websocket` protocol.
* Create the `nyc-open-data` index.
* Create the `yellow-taxi` collection (within the `nyc-open-data` index).

## Create your first document

> Now you successfully connected to your Kuzzle server with the Go SDK and have an index and a collection, it's time to manipulate data. Kuzzle use `indexes` that contain `collections`.

Reviewer: ```suggestion
Now you successfully connected to your Kuzzle server with the Go SDK and have an index and a collection, it's time to manipulate data. Kuzzle uses `indexes` that contain `collections`.
```

Those `collections` store `documents`.
Create a `document.go` file in the playground and add this code:

[snippet=document]

As you did before, run your program:

```bash
$ go run document.go
Connected!
New document added to yellow-taxi collection!
```

You can perform other actions such as [delete]({{ site_base_path }}sdk-reference/go/1/document/delete),
[replace]({{ site_base_path }}sdk-reference/go/1/document/replace) or [search]({{ site_base_path }}sdk-reference/go/1/document/search) documents. There are also other way to manipulate data in Kuzzle
like our [Admin Console]({{ site_base_path }}guide/2/essentials/installing-console/), the [Kuzzle HTTP API]({{ site_base_path }}api/1/essentials/connecting-to-kuzzle/) or your [own protocol](protocols/1/essentials/getting-started/).

> Now, you know how to:

Reviewer: ```suggestion
Now you know how to:
```

---
layout: sdk.html.hbs
algolia: true
title: Getting started
description: Getting started
order: 100
separator: essentials
---

# Getting Started

In this tutorial you will learn how to install the Kuzzle **Go SDK**.
We will walk you through creating scripts that can **store** documents in Kuzzle and subscribe to **notification** for each new document created.

<div class="alert alert-success">
Before proceeding, please make sure your system has these programs installed:

* **Go** version 1.9 or higher (you can find instructions <a href="https://golang.org/doc/install">here</a>).
* A running Kuzzle server (you can find instructions <a href="guide/1/essentials/installing-kuzzle/">here</a>).
</div>

## Installation

You can easily install our Go SDK:

```bash
$ go get github.com/kuzzleio/sdk-go
```

This will fetch the SDK and install it in your `GOPATH`.

## First connection

Initialize a new Go project as described in [Go Documentation](https://golang.org/doc/code.html#Command).
Then create a `init.go` file and start by adding the code below:

[snippet=init]

This program will initialize our Kuzzle creating an index and a collection
we'll use in the next section.

> Run this little bunch of code:

Reviewer 2: ```suggestion
Run the program with the following command:
```

Those `collections` store `documents`.
Create a `document.go` file in the playground and add this code:

[snippet=document]

As you did before, run your program:

```bash
$ go run document.go
Connected!
New document added to yellow-taxi collection!
```

You can perform other actions such as [delete]({{ site_base_path }}sdk-reference/go/1/document/delete),
[replace]({{ site_base_path }}sdk-reference/go/1/document/replace) or [search]({{ site_base_path }}sdk-reference/go/1/document/search) documents. There are also other way to manipulate data in Kuzzle
like our [Admin Console]({{ site_base_path }}guide/2/essentials/installing-console/), the [Kuzzle HTTP API]({{ site_base_path }}api/1/essentials/connecting-to-kuzzle/) or your [own protocol](protocols/1/essentials/getting-started/).

Now you know how to:
* Store data in your Kuzzle server with documents and access it.

## Subscribe to realtime document notifications (pub/sub)

Time to use realtime with Kuzzle. Create a new file `realtime.go` containing:

[snippet=realtime]

> This program subscribe to `yellow-taxi` collection. For each new document filter matching, we'll receive notification.

Reviewer 3: ```suggestion
This program subscribes to changes made to documents with a `license` field set to `B`, within the `yellow-taxi` collection. Whenever a document matching the provided filters changes, a new notification is received from Kuzzle.
```

Run your program:

```bash
$ go run realtime.go
Connected!
Successfully subscribing!
New document added to yellow-taxi collection!
Driver John born on 1995-11-27 got a B license.
```

Now, you know how to:

> * Create realtime filters.

Reviewer 3: ```suggestion
* Create realtime filters
```

> * Subscribe to notifications.

Reviewer 3: ```suggestion
* Subscribe to notifications
```

<div class="alert alert-info">

> If you want to go deeper in the SDK, source code is available on <a href="https://github.com/kuzzleio/sdk-go">Github</a>.

Reviewer 3: Shouldn't we instruct to browse the current documentation instead?

Author: Yeah you're right. I'll change this one. :+1:

Author: We are already on the SDK documentation so we could remove this block.

</div>

<div class="alert alert-info">
Having trouble? Get in touch with us on <a href="https://gitter.im/kuzzleio/kuzzle">Gitter!</a> We're happy to help.
</div>

## Where do we go from here?

> Now that you're more familiar with Kuzzle Go SDK, dive even deeper to learn how to leverage its full capabilities:

Reviewer 3: ```suggestion
Now that you're more familiar with the Go SDK, you can dive even deeper to learn how to leverage its full capabilities:
```

> * take a look at the <a href="{{ site_base_path }}sdk-reference/js/6">SDK Reference</a>

Reviewer 3: This link actually refers to... the Javascript Getting Started page. Fixing the link to make it point to the Go SDK directory won't be much better, as this will make it point to the current page.

```suggestion
* discover what this SDK has to offer by browsing other sections of this documentation
```

* learn how to use <a href="{{ site_base_path }}koncorde/1">Koncorde</a> to create incredibly fine-grained and blazing-fast subscriptions

> * follow our guide to learn how to implement <a href="{{ site_base_path }}guide/1/essentials/user-authentication/#local-strategy">basic authentication</a>

Reviewer 3: ```suggestion
* follow our guide to learn how to perform <a href="{{ site_base_path }}guide/1/essentials/user-authentication/#local-strategy">basic authentication</a>
```

> * follow our guide to learn how to implement <a href="{{ site_base_path }}guide/1/essentials/security/">manage users and setup fine-grained access control</a>

Reviewer 3: ```suggestion
* follow our guide to learn how to <a href="{{ site_base_path }}guide/1/essentials/security/">manage users and how to set up fine-grained access control</a>
```
